[JENKINS-75378] Adding a CLI command listener

apuig opened the pull request:
See [JENKINS-75378](https://issues.jenkins.io/browse/JENKINS-75378).

The default implementation mimics the current logs in CLICommand.

### Testing done

Executed commands using cli.jar using different transports

### Proposed changelog entries

- Add a new ExtensionPoint to receive events from CLI executions.
- Move logs from CLICommand to a new default listener.
- Command implementations using CLIMethod now log under the same conditions as CLICommand.

### Proposed changelog category

/label rfe


### Proposed upgrade guidelines

N/A

<!-- Comment:
Leave the proposed upgrade guidelines in the pull request with the "N/A" value if no upgrade guidelines are needed.
The changelog generator relies on the presence of the upgrade guidelines section as part of its data extraction process.
-->

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

apuig committed:
[JENKINS-75378] cli command listener

apuig commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -253,38 +256,60 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
-            LOGGER.log(Level.FINE, "Invoking CLI command {0}, with {1} arguments, as user {2}.",
-                    new Object[] {getName(), args.size(), auth.getName()});
+
+            Listeners.notify(
```
`jenkins.util.Listeners` looks like a good option here. `asSystem=false`, the proposed interface pass along user authentication

daniel-beck replied:
> `asSystem=false`, the proposed interface pass along user authentication

Could you explain this in detail? Seems wrong to not impersonate SYSTEM in the listener if you pass along the authentication anyway.

apuig replied:
perhaps I misinterpreted the javadoc on Listeners#asSystem.  Changing to asSystem=true

apuig commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Called before.
+     *
+     * @param correlationId This value is used to correlate this command event to other, related command events.
```
as in `jenkins.util.ScriptListener`, an implementation could use it to calculate time or encode a spanID.
String is good enough? A `startTime` could make some things easier?

apuig commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,100 @@
+     * @param command The executed command.
+     * @param argsSize Number of arguments passed to the command.
+     * @param auth Authenticated user executing the command
+     * @param expected True when caller is handling the error to return an specific exitCode.
```
This was to keep the exact same behavior in the logs: `FINE` vs `WARNING`.
Another option will be checking the type of the exception and discard the ones catched on CLICommand (less maintainable?) -- or use WARNING only for unchecked exceptions.
Not a big fan of boolean parameters in method signature, we may want 2 different methods.

apuig commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,100 @@
+     * @param opaqueId ID used on the command output to indicate an authentication error.
+     * @param t Authentication error.
+     * */
+    default void onLoginFailed(
```
The current implementation have this specific case where it generates an ID to correlate command output and logs. Another option is to use the general `onError` (with `auth=null`), and keep the specific log in CLICommand

apuig commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @param correlationId This value is used to correlate this command event to other, related command events.
+     * @param command The command to be executed.
+     * @param argsSize Number of arguments passed to the command.
```
Giving the full content of the arguments to the listeners implementation could lead to some security concern, for sure there is a proper way to filter-out sensitive data, but not sure if needed for this kind of listener

aneveux replied:
I agree that giving the full content of arguments could be risky. I am not sure whether sharing the argsSize is interesting or not though :shrug:

apuig commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,100 @@
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.ExtensionPoint;
+import hudson.cli.CLICommand;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.core.Authentication;
```
Perhaps its fine to use `String username` and avoid those imports

apuig commented on test/src/test/java/hudson/cli/CliListenerTest.java:
```diff
@@ -0,0 +1,96 @@
+import org.jvnet.hudson.test.LoggerRule;
+import org.jvnet.hudson.test.MockAuthorizationStrategy;
+
+public class CliListenerTest {
```
the tests are about the Default implementation (ala ScriptListenerTest), perhaps its better to introduce a custom impl as `@TestExtension`  ?

aneveux replied:
Looks good to me that way TBH.

jeromepochat commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -293,16 +318,17 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
+                        correlationId, getName(), args.size(), getTransportAuthentication2(), true, e));
 
-        logAndPrintError(e, e.getMessage(), logMessage, Level.FINE);
+        printError(e.getMessage());
```
**question:** Should the logging be extracted as listener implementation notified as other listeners?

apuig replied:
I think I don't understand the question. `printError` write to the command stderr stream

jeromepochat replied:
Sorry for unclear comment 😞

My concert was about having `notifyError` method that just notify listeners without any other behavior.

The initial `logFailedCommandAndPrintExceptionErrorMessage` method could me a listener implementation, just logging the event received.

IMO, this would align the proposal with common [Observer pattern](https://en.wikipedia.org/wiki/Observer_pattern) (aka as "Listener" in Java eco-system).

jeromepochat commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Callback around {@link CLICommand#run()}, each execution generates a new `correlationId` in order to group related events.
+ */
+public interface CliListener extends ExtensionPoint {
```
**todo:** Rename to `CLIListener` to align with other CLI related types.

apuig replied:
There are counter-examples:
CliAuthenticator, CliCrumbExclusio, CliLink ...

apuig replied:
the competition is disputed xD
```
find . | grep Cli | wc -l
11
find . | grep CLI | wc -l
11
```

Wadeck replied:
Agree about the current situation, not great consistency :)

I am not aware of any rules in Jenkins related to the abbreviation naming (i.e. following Java standards vs going against).

I updated the PR description to only contain the CliXxx format.

daniel-beck replied:
> CliAuthenticator

Deprecated

> CliCrumbExclusion

`DoNotUse.class`

> CliLink

Not intended for use outside core.

There's also the long-deprecated `CliTransportAuthenticator`.

OTOH, anything that anyone would refer to outside core is `CLI`: `CLICommand`, `CLIResolver`, `CLIMethod`. So for consistency with that, I suggest the types in this PR be renamed from `Cli` to `CLI`.

apuig replied:
now using `jenkins.cli.*CLI*`

jeromepochat commented on core/src/main/java/hudson/cli/listeners/DefaultCliListener.java:
```diff
@@ -0,0 +1,76 @@
+ * Basic default implementation of {@link CliListener} that just logs.
+ */
+@Extension
+public class DefaultCliListener implements CliListener {
```
**todo:** Rename to `DefaultCLIListener` to align with other CLI related types and similar [`jenkins.util.DefaultScriptListener`](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/util/DefaultScriptListener.java).

apuig committed:
fix checkstyle

apuig committed:
impersonate as SYSTEM in listeners

apuig committed:
introduce `fire` methods on the interface

apuig committed:
@since in interface. @Restricted default impl

apuig committed:
Simplify CliListener interface.
Add CliListener to CLIMethods

aneveux reviewed (approved):
Looks very good to me, thanks a lot! :heart:

apuig commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -253,56 +255,59 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
+        } else if (e instanceof AccessDeniedException) {
+            exitCode = 6;
+            printError(e.getMessage());
+        } else if (e instanceof BadCredentialsException) {
```
Looks like `BadCredentialsException` is not longer possible inside `CLICommand#main`, note `{set/get}TransportAuth2`:

* for HTTP or WebSocket the BadCredential is thrown/handled in the filters. When using command line `-auth user:pass` basic auth : BasicHeaderAuthenticator / AbstractUserAuth.

* for SSH auth errors (not a BadCredentials) is also handled before executing the command (PublicKeyAuthenticatorImpl / UserAuthNamedFactory).

https://github.com/jenkinsci/jenkins/commit/b1803a9539ec773e91a9fd126a7ae3c17c98794b I can't find CliAuthenticationTest, may not apply anymore.

jeromepochat commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,145 @@
+     * @param auth Authenticated user performing the execution.
+     *
+     * */
+    default void onExecution(
+            @NonNull String correlationId, @NonNull String command, int argsSize, @CheckForNull Authentication auth) {}
```
**question:** Does it worth to introduce a parameter object (e.g. `CliContext`) that groups together information common to all listener methods?
* `correlationId`
* `command`
* `argSize`
* `auth`

This could simplify method signatures.
* `onExecution(context)`
* `onCompleted(context, exitCode)`
* `onError(context, exitCode, t)`
* `fireExecution(context)`
* `fireCompleted(context, exitCode)`
* `fireError(context, t)`
* ...

WDYT?

apuig replied:
excellent suggestion!, working on it. It could also encapsulate the need of `correlationId`

apuig replied:
https://github.com/jenkinsci/jenkins/pull/10382/commits/b8f360aad9bb218f818ad35b76af3bc8d4b16c27

jeromepochat commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -239,6 +238,9 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
         this.locale = locale;
         CmdLineParser p = getCmdLineParser();
 
+        final String correlationId = UUID.randomUUID().toString();
```
**suggestion:** The introduction of `correlationId` common to all events could be highlighted in the PR description.

apuig replied:
At this point `correlationId` is only used in `BadCredentialsException` handling. (where the original randomUUID comes form)

apuig committed:
introduce hudson.cli.listeners.CliContext

jeromepochat commented on core/src/main/java/hudson/cli/listeners/CliContext.java:
```diff
@@ -0,0 +1,62 @@
+package hudson.cli.listeners;
```
**todo:** File header is missing.

jeromepochat commented:
Looks good, thanks @apuig !

apuig committed:
missing header in CliContext

Wadeck commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -239,70 +238,75 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
+            printError(e.getMessage());
+        } else if (e instanceof BadCredentialsException) {
+            exitCode = 7;
+            // to the caller, we can't reveal whether the user didn't exist or the password didn't match.
+            // do that to the server log instead
```
💡 this kind of information is no longer at the correct location I think, better to be included in DefaultCliListener

Wadeck commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -239,70 +238,75 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
         this.locale = locale;
         CmdLineParser p = getCmdLineParser();
 
+        final CliContext context = new CliContext(getName(), args.size(), getTransportAuthentication2());
```
```suggestion
        CliContext context = new CliContext(getName(), args.size(), getTransportAuthentication2());
```

💡 The usage of `final` is not consistent with the surrounding code. I would suggest to keep the same style in place as much as possible, except if there is a bug. Here it's a matter of personal opinion.

Same applies to another occurrence below

Wadeck replied:
The context approach is great for future improvements, to not require to touch every callers. 👍

Wadeck commented on core/src/main/java/hudson/cli/declarative/CLIRegisterer.java:
```diff
@@ -232,47 +230,13 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
-                                logAndPrintError(e, errorMsg, errorMsg, Level.WARNING);
-                                Functions.printStackTrace(e, stderr);
-                                return 1;
+                            int exitCode = handleException(e, context, parser);
+                            CliListener.fireError(context, exitCode, e);
+                            return exitCode;
```
```suggestion
                                int exitCode = handleException(e, context, parser);
                                CliListener.fireError(context, exitCode, e);
                                return exitCode;
```

❌ Wrong identation

I wonder if this could pass the auto style check (and if yes, why :))

apuig replied:
:see_no_evil: `mvn verify` didn't spot this

Wadeck commented on core/src/main/java/hudson/cli/listeners/CliContext.java:
```diff
@@ -0,0 +1,86 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2025
```
```suggestion
 * Copyright (c) 2025, CloudBees, Inc.
```

Should be applied to all new files.
With or without the (c), not totally sure due to existing occurrences.

Wadeck commented on core/src/main/java/hudson/cli/listeners/CliContext.java:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Holds information of a command execution. Same instance is used to all {@link CliListener} invocations.
+ * Use `correlationId` in order to group related events.
```
```suggestion
 * Use `correlationId` in order to group related events to the same command.
```

💡 Is the current message clear for any readers?

Wadeck commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,92 @@
+import jenkins.util.Listeners;
+
+/**
+ * Callback around {@link CLICommand#run()}
```
nit: Not sure it's a good description of what the class is responsible for, way too generic I think.

Wadeck reviewed (commented):
Pretty good concept, thanks for the proposal Albert 👍 
Some minor comments here and there.

Not manually tested.

apuig committed:
pr comments

kmartens27 commented:
Hi @apuig, would this need to be added to the jenkins.io documentation, and if so is there any additional content/context beyond what is described in this and the JIRA tickets?

jglick commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,92 @@
+     * Invoked on execution failure.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode `run` returned exit code.
```
It seems unnecessary to pass this parameter to the listener. If anyone really cared it could be inferred by the same logic as in `CLICommand` but generally a listener is going to want to report any error at the Java level, not by using a Unix-like exit code value.

apuig replied:
ATM exit code is used to respect the current log levels. Notably unhanded exceptions and BadCredentials.
We can indeed check the same kind of Exceptions as in CLICommand#handleException, but  exit codes handled by jenkins are properly documented on CLICommand#main, perhaps more maintainable ?

apuig replied:
exit code removed in the signature with the exception

jglick commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,92 @@
+     * */
+    default void onError(@NonNull CliContext context, int exitCode, @NonNull Throwable t) {}
+
+    /**
+     * Fires the {@link #onExecution} event.
+     *
+     * @param context Information about the command being executed
+     * */
+    static void fireExecution(@NonNull CliContext context) {
+        Listeners.notify(CliListener.class, true, listener -> listener.onExecution(context));
+    }
+
+    /**
+     * Fires the {@link #onCompleted} event.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode `run` returned exit code.
+     * */
+    static void fireCompleted(@NonNull CliContext context, int exitCode) {
+        Listeners.notify(CliListener.class, true, listener -> listener.onCompleted(context, exitCode));
+    }
+
+    /**
+     * Fires the {@link #onError} event.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode `run` returned exit code.
+     * @param t Any error during the execution of the command.
+     * */
+    static void fireError(@NonNull CliContext context, int exitCode, @NonNull Throwable t) {
+        Listeners.notify(CliListener.class, true, listener -> listener.onError(context, exitCode, t));
+    }
```
Or these could all be inlined as there is no valuable code here.

jglick commented on test/src/test/java/hudson/cli/CliListenerTest.java:
```diff
@@ -0,0 +1,168 @@
+import org.jvnet.hudson.test.MockAuthorizationStrategy;
+import org.jvnet.hudson.test.TestExtension;
+
+public class CliListenerTest {
```
Or `DefaultCliListenerTest`, since that is what it is (directly) testing.

jglick commented on core/src/main/java/hudson/cli/listeners/CliContext.java:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Holds information of a command execution. Same instance is used to all {@link CliListener} invocations.
+ * Use `correlationId` in order to group related events to the same command.
```
```suggestion
 * Use {@code correlationId} in order to group related events to the same command.
```
alas

apuig commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,92 @@
+     * Invoked after successful execution.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode `run` returned exit code.
```
Implementations of `CLICommand#run` can return other values to indicate custom errors. 

I try to avoid `onSuccess` because of this (but the javadoc comment still mention _success_ :S), the idea is we call this if the command _returned_ instead of throwing an exception.
I agree `onCompleted` vs `onError` its not clear enough, but can't find a better option (`onException`?)

apuig committed:
pr comments:

clarify interface about onComplete not always success

remove exitCode from onException

remove fire methods

args in CliContext

rename CliListenerTest

apuig commented:
please note: the [failing test](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-10382/10/realtimeTestReport-130/junit/hudson.cli/DisconnectNodeCommandTest/linux_jdk17___Linux___JDK_17___Build___Test___disconnectNodeManyShouldSucceed/)
> hudson.cli.DisconnectNodeCommandTest.disconnectNodeManyShouldSucceed

reference CLI, but master also experience problems with [this test class](https://ci.jenkins.io/job/Core/job/jenkins/job/master/7139/testReport/)

krisstern committed:
Merge branch 'master' into JENKINS-75378-cli-listener

krisstern commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -242,70 +242,73 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
 
-    private void logAndPrintError(Throwable e, String errorMessage, String logMessage, Level logLevel) {
-        LOGGER.log(logLevel, logMessage, e);
+
```
```suggestion
```

krisstern reviewed (approved):
Makes sense to me

krisstern committed:
Update core/src/main/java/hudson/cli/CLICommand.java

krisstern commented:
/label ready-for-merge

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

alecharp reviewed (approved):
Looks good to me.

daniel-beck commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -242,70 +242,72 @@ public int main(List<String> args, Locale locale, InputStream stdin, PrintStream
             old = sc.getAuthentication();
 
-            sc.setAuthentication(auth = getTransportAuthentication2());
+            sc.setAuthentication(getTransportAuthentication2());
```
Not: Could still assign `#getTransportAuthentication` to a variable instead of calling it twice.

daniel-beck commented on core/src/main/java/hudson/cli/listeners/CliListener.java:
```diff
@@ -0,0 +1,60 @@
+     * @param context Information about the command being executed
+     * @param t Any error during the execution of the command.
+     * */
+    default void onException(@NonNull CliContext context, @NonNull Throwable t) {}
```
Misleading name if it takes a `Throwable`? `#onError` maybe?

apuig replied:
I renamed `onError` to `onException` to better emphasize that `onComplete` also handles cases where `run` returns an error code.
https://github.com/jenkinsci/jenkins/pull/10382#discussion_r1989712724
I'm open to suggestions, but I prefer to avoid using `onError`.

daniel-beck replied:
JUnit has failures (unsatisfied assertions) and errors (unexpected exceptions thrown), perhaps we could use one term to refer to "did not succeed" and use the other for "this threw up"?

daniel-beck commented on core/src/main/java/hudson/cli/listeners/CliContext.java:
```diff
@@ -0,0 +1,88 @@
+    private final Authentication auth;
+
+    /**
+     * @param command The command being executed.
```
Have you considered passing the `CLICommand` as argument, rather than just the name?

apuig replied:
I considered it but wasn't fully convinced. It seems safe to pass CLICommand (e.g., it doesn't expose the I/O streams for manipulation), but there might be (?) some CLICommand implementations with public methods that could be problematic if a listener were able to use them.
(*) If an implementation requires it, it can use the `CLICommand.getCurrent` thread-local. However, it seems that some fixes may be needed for the SSH transport.

daniel-beck commented on test/src/test/java/hudson/cli/DefaultCliListenerTest.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    @Test
+    public void commandExecutionUnexpectedErrorIsLogged() throws Exception {
```
This test has no corresponding `actionExecutionUnexpectedErrorIsLogged`, why?

apuig replied:
you got me. I was unable to create a test implementation for a `@CLIMethod`, possibly due to a limitation in the CLICommandInvoker utility. Additionally, I couldn't find any similar tests.
Do you have any suggestions on how to trigger an exception in an existing default `@CLIMethod`?

daniel-beck commented on core/src/main/java/hudson/cli/listeners/CliContext.java:
New types should be in `jenkins.*` unless there are reasons not to. The same applies to other new classes in this PR.

apuig committed:
move new classes to jenkins.cli package

apuig committed:
rename new classes to capitalize CLI

apuig committed:
single call to getTransportAuthentication2

daniel-beck reviewed (approved):
Looks great 👍 Thanks!

Would be nice to have the missing test, and perhaps checking the `onException` name, but not blocking. I'll leave it to @apuig whether to address those.

apuig committed:
test CLIMethod unexpected error

apuig committed:
rename to `onThrowable`

apuig committed:
Merge branch 'master' into JENKINS-75378-cli-listener

apuig commented:
Thanks for the review  @daniel-beck . With your hint, I was able to write the missing test.
Regarding naming, we're going with `onThrowable`, for consistency with both the parameter type and when it's called.

The current failing test [`PasswordParameterDefinitionTest.defaultValueAlwaysAvailable`](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-10382/19/) is also present in master, so I belive it's unrelated to this PR.

Would it be possible to have another review to put back the `ready-for-merge` lable?

daniel-beck commented:
I'm going to start the countdown here, but welcome further feedback on the renamed method before we're cementing it forever 😉

A1exKH reviewed (approved):
LGTM.

jglick commented on core/src/main/java/jenkins/cli/listeners/CLIListener.java:
```diff
@@ -0,0 +1,60 @@
+     * Invoked after command execution.
+     *
+     * @param context Information about the command being executed.
+     * @param exitCode Exit code returned by the implementation of {@link CLICommand#run()}.
```
How would this not always be zero?

apuig replied:
yeah, that was the tricky part on the naming, indeed `onCompleted`  will handle controlled exit codes from implementations

jglick replied:
The Javadoc could probably be clarified to emphasize that either `onCompleted` or `onThrowable` is called.

apuig replied:
HTH https://github.com/jenkinsci/jenkins/pull/10473
